notebooks/hello-universe/Interpreting_CNNs/Interpreting_CNNs.ipynb

```diff
@@ -0,0 +1,803 @@
+    "# dimensions\n",
+    "(w, h) = (X_test.shape[2], X_test.shape[1])\n",
+    "heatmap = cv2.resize(cam.numpy(), (w, h))\n",
+    "\n",
+    "# normalize the heatmap such that all values lie in the range\n",
+    "# [0, 1], scale the resulting values to the range [0, 255],\n",
+    "# and then convert to an unsigned 8-bit integer"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "3ad137fa-2d40-48f6-b2c4-1dc50b106eb9",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "y_pred = cnn.predict(img)\n",
+    "\n",
+    "# Plot the results next to the original image\n",
+    "fig, axes = plt.subplots(1, 3, figsize=(14, 5))\n",
+    "axes[0].imshow(orig_img)\n",
+    "axes[0].set_title(\"orig_img\")\n",
+    "axes[1].imshow(scaled_img)\n",
+    "axes[1].set_title(\"scaled_img\")\n",
+    "i = axes[2].imshow(heatmap, cmap=\"jet\", alpha=0.8)\n",
+    "fig.colorbar(i)\n",
+    "axes[2].set_title(\"heat_map\")\n",
+    "fig.suptitle(\"prediction_val=\" + str(y_pred))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a8992e02-d1d2-4314-8849-e4db7932469e",
+   "metadata": {},
+   "source": [
+    "**Image caption**: The above image contains three panels in a horizontal row. The first panel shows the original image of a merginig galaxy candidate, the second panel shows a logarithmically-scaled version of the original image, and the third panel shows a heat map of the grad-cam map."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "697a59ff-53b3-4353-804c-5061281d5b3a",
+   "metadata": {},
+   "source": [
+    "**The result is very similar** to our saliency map from backpropagation, with the image being coarser but also showing bot the top and bottom of the galaxy. We can also play around with the selected layer to calculate the output at different steps in the model and see how activation changes throughout the model."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "27909148-c421-48b6-9ecb-a47f56a5dd61",
+   "metadata": {
+    "tags": []
+   },
+   "source": [
+    "## 3. RISE Algorithm\n",
+    "The RISE (Randomized Input Sampling for Explanation) Algorithm is another interpretation technique for calculating saliency maps. Instead of calculating gradients from within the model, the RISE implementation works by covering up pieces of the input image, running it through the model, and calculating the average activation in order to determine what parts of the image are most important. As such, this method does not require any access to the inner workings of the model. The algorithm first generates a sequence random of binary grids (called masks), which are placed onto the image. Everything not covered by the mask is removed by multiplying the images together, and the resultant activations are averaged to get our final heatmap. A full description of the algorithm and its variations is provided by <a href='https://arxiv.org/abs/1806.07421'>Petsiuk et al. 2018</a>. \n",
```

Comment:
```suggestion
    "The RISE (Randomized Input Sampling for Explanation) Algorithm is another interpretation technique for calculating saliency maps. Instead of calculating gradients from within the model, the RISE implementation works by covering up pieces of the input image, running it through the model, and calculating the average activation in order to determine what parts of the image are most important. As such, this method does not require any access to the inner workings of the model. The algorithm first generates a random sequence of binary grids (called masks), which are placed onto the image. Everything not covered by the mask is removed by multiplying the images together, and the resultant activations are averaged to get our final heatmap. A full description of the algorithm and its variations is provided by <a href='https://arxiv.org/abs/1806.07421'>Petsiuk et al. 2018</a>. \n",
```